Add examples

preiter93 opened the pull request:
Hi,

I think it would be nice to add a few more examples to expose the most important functions
more prominently. Especially the mpi example could be usefull for some users.

mulimoen commented on examples/read_write_ndarray_mpi.rs:
```diff
@@ -0,0 +1,137 @@
+where
+    A: hdf5::H5Type,
+    D: ndarray::Dimension,
+    T: AsRef<[ndarray::SliceInfoElem]>,
```
Try some variant of `TryInto<Selection>` here for idiomatic code

mulimoen commented on examples/read_write_ndarray_mpi.rs:
```diff
@@ -0,0 +1,137 @@
+    T: AsRef<[ndarray::SliceInfoElem]>,
+{
+    // Open file
+    let file = hdf5::FileBuilder::new().with_fapl(|p| p.mpio(comm, None)).open(filename)?;
```
Move the opening of the file outside of the `read` function

mulimoen commented on examples/read_write_ndarray_mpi.rs:
```diff
@@ -0,0 +1,137 @@
+    T: AsRef<[ndarray::SliceInfoElem]>,
+    Sh: Into<hdf5::Extents>,
+{
+    use std::convert::TryFrom;
```
Move to top of file

mulimoen commented on examples/read_write_ndarray_mpi.rs:
```diff
@@ -0,0 +1,137 @@
+    );
+
+    // Create file with mpi access
+    let file = hdf5::FileBuilder::new().with_fapl(|p| p.mpio(comm, None)).append(filename)?;
```
Move outside of the function

mulimoen commented on examples/read_write_ndarray_mpi.rs:
```diff
@@ -0,0 +1,137 @@
+        // Overwrite
+        Ok(dset) => dset,
+        // Create new dataset
+        std::prelude::v1::Err(..) => {
```
Drop `std::prelude`, `Err(_)` should match here

mulimoen replied:
Or maybe just unwrap/? if the file creation is moved outside of the function

mulimoen commented on examples/read_write_ndarray_mpi.rs:
```diff
@@ -0,0 +1,137 @@
+    if initialized == 0 {
+        unsafe { mpi_sys::MPI_Init(ptr::null_mut(), ptr::null_mut()) };
+    }
+    let world_comm = unsafe { mpi_sys::RSMPI_COMM_WORLD };
```
Unnecssary `unsafe`?

preiter93 replied:
seems necessary (use of mutable static)

mulimoen commented on examples/read_write_ndarray_mpi.rs:
```diff
@@ -0,0 +1,137 @@
+    }
+    let world_comm = unsafe { mpi_sys::RSMPI_COMM_WORLD };
+    let mut rank: c_int = 1;
+    unsafe { mpi_sys::MPI_Comm_rank(world_comm, &mut rank) };
```
You should assert `world-size == 2`

mulimoen reviewed (commented):
Great to see another example! It would be great if these two examples could be merged into one; where the MPI isage is shown side-by-side the "normal" usage. I think `mpi` is fairly abstracted away and the differences could be made fairly minor

preiter93 commented:
Thank you for the review!

I have removed the previous examples and added two new ones, one for serial read/write
and the other for mpi read/write. I think it is better to separate serial and parallel.

I'm not quite sure when to use file.close(). For the mpi example it will fail, so I'll let the close step out.

And feel free to edit the examples, I have only superficial knowledge with this crate.

mulimoen commented:
@preiter93 I had a small go over here: https://github.com/mulimoen/hdf5-rust/pull/15, do these modifications make sense in how you want to examplify the crate?

preiter93 commented:
@mulimoen Yes looks very nice now. Just made two tiny changes:

- #[cfg(mpio)] -> #[cfg(feature = "mpio")]
- Remove lifetime of read
